### ISSUE TITLE
Cawbird accessibility

### ISSUE THREAD

**Issue by serrebi** [IBBoard/cawbird#143]
I am an orca(gnome-orca) user. Cawbird needs some work in the list of tweets. Right now it only reads the name but if I use object nav(some of the time) I can read the tweet text, click on links etc.
Here are some links that might help you.
https://accessibility.linuxfoundation.org/a11yspecs/atspi/adoc/ADOC_ATSPI.html
https://accessibility.linuxfoundation.org/a11yspecs/atspi/adoc/a11y-dom-apis_20080324.html

**Comment by IBBoard** [comment]
Thanks, I'd seen a couple of comments on Twitter about accessibility, but I'd never managed to get a response from people about what screen reader they were using and what their problem was (what wasn't reading, what parts of the UI were hard to access, or whatever)

**Comment by serrebi** [comment]
Hey: I'm one of those people. Apologies, I missed your message asking me about the screen reader thing. Sorry about that. I'm glad you liked my feedback. The only screen reader there is for the GUI on Linux is gnome-Orca. So that's all blind Linux users will be using, unless something changes.

**Comment by IBBoard** [comment]
Just installed Orca and checking some of this now. I think my first step is going to be working out what it is and isn't reading, and why!

It appears to read in the order that the GTK Inspector shows the widgets (so RT labels are listed first and read first) but it doesn't appear to be consistent in what it reads. So far I've found:

* Most RTs read the retweeter's name, the replying to line (if it exists), the time and the retweeted accounts handle (but not the name)
* One RT reads the the retweeted account's name as well
* One RT just reads the retweeter's name and nothing else!

And a quick guess is that the tweets aren't being read because they're in a potentially scrollable widget, which is necessary to fix wide tweets forcing the app to go too wide (d98f390d88f21fb37f49d0d5c69b100e9be80c82)

If there's any other specific things to fix then please tell me.

**Comment by IBBoard** [comment]
I've been trying to work out what Orca is doing with some of these widgets and why it is or isn't reading things. And it isn't clear.

I've tried to add accessible descriptions to some icons (such as the "Retweet" icon) but haven't yet made Orca read one.

Everything looks okay in Accerciser, though.

Someone somewhere knows how this works, but it isn't me and I've not yet worked it out!

**Comment by IBBoard** [comment]
It's a tiny thing so far, and there's already several other things that I know needs doing, but that new branch at least makes Cawbird read tweets properly!

Now to get the accessible descriptions working…

**Comment by IBBoard** [comment]
Quick question on "best" accessibility behaviour:

On the search page, when you hit "Enter" to run the search, is it best for the search results to grab focus when the search completes or is it okay to just press "down" to get to the results?

I'm just trying to work out how you'd indicate that there are now results available after the search is run!

**Comment by IBBoard** [comment]
And another thought/question: How accessible is the "right-click to display RT/Like/Quote/etc buttons" behaviour?

**Comment by IBBoard** [comment]
And another thought/question: Is it better to have things like people's names as push buttons or links?

Having it read "IBBoard push button at ibboard 17 minutes ago blah blah blah… Professor Brian Cox push button at Prof Brian Cox 20 minutes ago…" seems like a lot of unnecessary "push button", but then I've realised that it might be less obvious that it's clickable to go to the profile.

Being a link then it's still in the tab order, though. And it's not like "push button" tells you what it *actually* does.

**Comment by IBBoard** [comment]
And another thought/question: With keyboard navigation, is the expectation that if "down" works then "up" always works?

Because currently things like pressing "Down" from the nav bar to the "Lists" view takes you to the list, but "up" gets stuck at the top of the list and won't go back.

And as a follow-up: If it is expected, should it take you back to the relevant button above (like going from a page to its tab)? So "up" from the lists goes to the "Lists" button in the nav, and "up" on the followers list on the profile page takes you to the followers button?

**Comment by IBBoard** [comment]
Still to do:
 * Read "quoted tweet" before reading quoted tweet
 * Read "Retweeted by" (currently shown as the "two arrows in a circle" symbol, and Orca just reads the name and handle)
 * Read "tweet in thread" or "reply" (little speech bubble next to the time to show it is a reply in a thread)
 * Read the profile when first loading it (there's currently no way to read location and bio)

And then I'm planning to do the following but need feedback:
 * Read "rewteeted"? (arrows by the time to show you retweeted it)
 * Read "liked"? (heart by the time to show you liked it)

Things I can't get working at the moment or need to investigate more:
 * Moving up out of lists of Twitter lists (but Shift+Tab works)
 * Load new tweets behaviour - I think I just had it stop mid-tweet and read the new tweet as soon as it appeared. This *might* be because it stole focus
 * Do we need a better way of doing the Like/RT instead of the hidden right-click interface?

And things that can be handled separately from this ticket:
 * #10 - Adding image descriptions when posting
 * #11 - Showing image descriptions on media

Did I miss anything?

**Comment by IBBoard** [comment]
Trying to research the problem. Looks like the "critical" log messages might be an *old* issue in the app!

baedert/corebird#174 is from 2014! Closed months later, but no details.

**Comment by IBBoard** [comment]
gdb output for debugging the criticals using `G_DEBUG=fatal-criticals gdb build/cawbird`:

```
(gdb) bt
#0  0x00007ffff7ed3942 in g_logv () at /usr/lib64/libglib-2.0.so.0
#1  0x00007ffff7ed3baf in g_log () at /usr/lib64/libglib-2.0.so.0
#2  0x00007ffff756a42a in atk_object_ref_state_set (accessible=0x0) at ../atk/atkobject.c:1019
#3  0x00007ffff777b098 in gtk_label_accessible_link_impl_ref_state_set (obj=0x5555571b1070) at a11y/gtklabelaccessible.c:105
#4  0x00007ffff68e5e5f in spi_atk_state_to_dbus_array (object=<optimized out>, array=0x7fffffffd040) at ../atk-adaptor/accessible-stateset.c:188
#5  0x00007ffff68ee727 in impl_GetState (bus=<optimized out>, message=<optimized out>, user_data=0x7fffffffcfa0)
    at ../atk-adaptor/adaptors/accessible-adaptor.c:448
#6  0x00007ffff68f0a61 in handle_other
    (pathstr=0x7fffe4018c18 "/org/a11y/atspi/accessible/692", member=0x7fffe4018c78 "GetState", iface=<optimized out>, path=0x555555c6d2b0, message=0x555556d0df10, bus=0x5555572286a0) at ../droute/droute.c:554
#7  0x00007ffff68f0a61 in handle_message (bus=0x5555572286a0, message=0x555556d0df10, user_data=0x555555c6d2b0) at ../droute/droute.c:601
#8  0x00007ffff5f24511 in dbus_connection_dispatch () at /usr/lib64/libdbus-1.so.3
#9  0x00007ffff5ee2635 in  () at /usr/lib64/libatspi.so.0
#10 0x00007ffff7ecc2b7 in g_main_context_dispatch () at /usr/lib64/libglib-2.0.so.0
#11 0x00007ffff7ecc638 in  () at /usr/lib64/libglib-2.0.so.0
#12 0x00007ffff7ecc6ef in g_main_context_iteration () at /usr/lib64/libglib-2.0.so.0
#13 0x00007ffff72f4845 in g_application_run () at /usr/lib64/libgio-2.0.so.0
#14 0x0000555555570e75 in _vala_main (args=0x7fffffffd528, args_length1=1) at ../src/main.vala:26
#15 0x0000555555570fbd in main (argc=1, argv=0x7fffffffd528) at ../src/main.vala:18
```
(You'll need various debug packages as well - atk, gtk, glib2, dbus)

So it seems to be a hyperlink that's causing the issue. Frame 3 is:
```
state_set = atk_object_ref_state_set (atk_object_get_parent (obj));
```
So the `atk_object_get_parent` call on the hyperlink returns a null pointer, which obviously fails the "is it an accessible object" check in frame 2. Don't know why yet, or which widget, but that's why.

**Comment by IBBoard** [comment]
I think we're about as far as we can get with the profile now. It should still read on mouse over (if you've got that enabled) but for now it'll just give the location (most of the time) and the bio text.

**Comment by IBBoard** [comment]
I've just made "Cawbird Unstable" builds available for anyone who wants to test the (hopefully) improved accessibility but who can't build their own version from Git. They include all of the changes from here and the image description changes from #10 and #11.

https://software.opensuse.org//download.html?project=home%3AIBBoard%3Acawbird-unstable&package=cawbird-unstable

**NOTE:** These are unstable development builds. They're whatever subset of extra changes have been committed to `git master` that I think I need the community to test. They may break things. They should only be used for testing. They shouldn't be relied upon for normal use. I'd recommend backing up `~/.config/cawbird` before running them. You have been warned!

(If they're not showing yet then give it a while - the packages are built, they're just not showing on that page yet)

**Comment by IBBoard** [comment]
Okay, I'm going to call it quits on this for now. I think we're in a much better position than we were. If anyone finds anything else that needs improving then please open a ticket and we'll look at it. We also need to keep in mind the future redesign in #177 and how that will affect things.
